### PR TITLE
Improve email validation flow in the console and cli

### LINKED ISSUE #1290: Improve email validation flow
#### Summary

We need to improve the email address validation flow.

#### Why do we need this?

It's not user friendly right now.

#### What is already there? What do you see now?

1. We have the `ContactInfoRegistry.Validate` RPC
2. We send the following email:

> Please confirm your email address for The Things Stack.
>
> Your email address will be used as contact for user "admin".
>
> Reference: XXX Confirmation Token: XXX

#### What is missing? What do you want to see?

- Neither the CLI nor the Console has functionality to call the Validate RPC
- The email needs to show a link for validation in the console
- The email needs to show the command for validation in the CLI

#### How do you propose to implement this?

- In the CLI it's straightforward, just add a command that takes the Ref and Token and calls the RPC
- In the Console it would be similar to how we did the password reset
- Same for the changes in the email template

#### Can you do this yourself and submit a Pull Request?

Probably faster if someone else does the Console part. I can do the CLI part if needed.

### PR DESCRIPTION
**Pull request by pgalic96**
#### Summary

Closes #1290 

#### Changes

- Add `Validate` view that makes an api request with url encoded parameters
- Extend validation email
- Add api endpoint in the oauth app
- Add cli command for validating contact info with reference and token params

#### Notes for Reviewers

- @kschiffer not sure what would be the best option to display once the request succeeds
- @htdvisser I moved the validate-email command under the users cli commands, since organizations, applications, etc also have contact info commands, is there a scenario in which we would use this `validate email` commands not only for user but for orgs for example? Also this command is not working yet, trying to find out what's the issue.
#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [ ] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by htdvisser** [comment]
Entities such as Applications, Gateways, Organizations and OAuth Clients can have contact information too. Maybe someone would want to have a public technical or abuse contact for an application or gateway

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 1]
```diff
@@ -26,7 +26,7 @@ import (
 	"go.thethings.network/lorawan-stack/pkg/errors"
 	"go.thethings.network/lorawan-stack/pkg/ttnpb"
 )
-
+var errMissingParam = errors.DefineInvalidArgument("missing_params", "missing token or reference")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/contact_info.go]: This doesn't look quite right:

- Spacing is off
- Not located "as close as possible to where used"

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 2]
```diff
@@ -178,3 +178,38 @@ func contactInfoCommands(entity string, getID func(cmd *cobra.Command, args []st
 	cmd.AddCommand(add, remove)
 	return cmd
 }
+
+func validateEmailCommand() *cobra.Command {
+	validateEmail := &cobra.Command{
+		Use:			"validate-email [reference] [confirmation-token]",
+		Short: 		"Validate account email",
+		RunE: func(cmd *cobra.Command, args []string) error {
+
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/contact_info.go]: Where does this newline come from?

File: cmd/ttn-lw-cli/commands/users.go [hunk 3]
```diff
@@ -358,5 +358,6 @@ func init() {
 	usersCommand.AddCommand(usersDeleteCommand)
 	usersContactInfoCommand.PersistentFlags().AddFlagSet(userIDFlags())
 	usersCommand.AddCommand(usersContactInfoCommand)
+	usersCommand.AddCommand(validateEmailCommand())
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/users.go]: Let's use a `validate` in `contact-info`

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 4]
```diff
@@ -109,6 +109,7 @@ var contactInfoFlags = util.FieldFlags(&ttnpb.ContactInfo{})
 var (
 	errContactInfoExists           = errors.DefineAlreadyExists("contact_info_exists", "contact info already exists")
 	errMatchingContactInfoNotFound = errors.DefineAlreadyExists("contact_info_not_found", "matching contact info not found")
+	errMissingParam                = errors.DefineInvalidArgument("missing_params", "missing token or reference")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/contact_info.go]: Create `errNoReference` and `errNoValidationToken` instead

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 5]
```diff
@@ -173,8 +174,36 @@ func contactInfoCommands(entity string, getID func(cmd *cobra.Command, args []st
 			return io.Write(os.Stdout, config.OutputFormat, updatedInfo)
 		},
 	}
+	validate := &cobra.Command{
+		Use:   "validate [reference] [confirmation-token]",
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/contact_info.go]: If you use `confirmation-token` here, the flag should also be called `confirmation-token`.

Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/contact_info.go]: Also, you're not actually checking the args in RunE below

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 6]
```diff
@@ -173,8 +174,36 @@ func contactInfoCommands(entity string, getID func(cmd *cobra.Command, args []st
...
+				ID:    reference,
+				Token: token,
+			})
+			if err != nil {
+				return err
+			}
+			return io.Write(os.Stdout, config.OutputFormat, res)
+		},
+	}
+	validate.Flags().String("reference", "", "")
```
Comment (htdvisser) [inline cmd/ttn-lw-cli/commands/contact_info.go]: Consider adding help text. `reference` by itself may not be clear enough

File: pkg/webui/oauth/views/app/index.js [hunk 7]
```diff
@@ -65,6 +66,7 @@ class OAuthApp extends React.PureComponent {
               <Route path="/forgot-password" component={ForgotPassword} />
               <Route path="/code" component={Code} />
               <Route path="/update-password" component={UpdatePassword} />
+              <Route path="/confirm-email" component={Validate} />
```
Comment (htdvisser) [inline pkg/webui/oauth/views/app/index.js]: Can we please be more consistent with "confirm" vs "validate"?

File: cmd/ttn-lw-cli/commands/contact_info.go (hunk 4 above)
Comment (johanstokking) [reply cmd/ttn-lw-cli/commands/contact_info.go]: Change error name accordingly

File: cmd/ttn-lw-cli/commands/contact_info.go (hunk 5 above)
Comment (johanstokking) [reply cmd/ttn-lw-cli/commands/contact_info.go]: I would just call it `token`; the context is clear enough here

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 8]
```diff
@@ -173,8 +174,36 @@ func contactInfoCommands(entity string, getID func(cmd *cobra.Command, args []st
...
+	validate.Flags().String("validation-token", "", "")
 	add.Flags().AddFlagSet(contactInfoFlags)
 	remove.Flags().AddFlagSet(contactInfoFlags)
-	cmd.AddCommand(add, remove)
+	cmd.AddCommand(add, remove, validate)
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/contact_info.go]: Do this per command to keep the diffs nice

File: pkg/webui/oauth/views/validate/index.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+          <Col sm={12} md={8} lg={5}>
+            <IntlHelmet title={m.confirmInfo} />
+            {error && (
+              <Notification
+                large
+                error={error}
+                title={m.confirmFail}
+                action={goToLogin}
+                actionMessage={m.goToLogin}
+                buttonIcon={'add_circle'}
```
Comment (kschiffer) [inline pkg/webui/oauth/views/validate/index.js]:
```suggestion
                buttonIcon={'error'}
```

File: pkg/webui/oauth/views/validate/index.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+              />
+            )}
+            {success && (
+              <Notification
+                large
+                success={success}
+                title={m.infoConfirmed}
+                action={goToLogin}
+                actionMessage={m.goToLogin}
+                buttonIcon={'add_circle'}
```
Comment (kschiffer) [inline pkg/webui/oauth/views/validate/index.js]:
```suggestion
                buttonIcon={'check_circle'}
```

File: pkg/webui/oauth/views/validate/index.js [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+import { connect } from 'react-redux'
+import IntlHelmet from '../../../lib/components/intl-helmet'
+import PropTypes from '../../../lib/prop-types'
+
+import Spinner from '../../../components/spinner'
+import Message from '../../../lib/components/message'
+import Notification from '../../../components/notification'
+import api from '../../api'
+import sharedMessages from '../../../lib/shared-messages'
+import style from '../create-account/create-account.styl'
```
Comment (kschiffer) [inline pkg/webui/oauth/views/validate/index.js]:
```suggestion

import style from '../create-account/create-account.styl'
```

File: pkg/webui/oauth/views/validate/index.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React from 'react'
+import { Container, Col, Row } from 'react-grid-system'
+import { defineMessages } from 'react-intl'
+import { push } from 'connected-react-router'
+import queryString from 'query-string'
+import { connect } from 'react-redux'
+import IntlHelmet from '../../../lib/components/intl-helmet'
```
Comment (kschiffer) [inline pkg/webui/oauth/views/validate/index.js]:
```suggestion

import IntlHelmet from '../../../lib/components/intl-helmet'
```

File: pkg/webui/oauth/views/validate/index.js [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+import PropTypes from '../../../lib/prop-types'
+
```
Comment (kschiffer) [inline pkg/webui/oauth/views/validate/index.js]:
```suggestion
```

File: cmd/ttn-lw-cli/commands/contact_info.go (hunk 6 above)
Comment (pgalic96) [reply cmd/ttn-lw-cli/commands/contact_info.go]: @htdvisser would "reference of your contact info validation (received in email)" suffice? assuming email is the only possible way of receiving it. or we can just exclude the part in braces.

Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/contact_info.go]: Or "that you received"

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 14]
```diff
@@ -109,6 +109,8 @@ var contactInfoFlags = util.FieldFlags(&ttnpb.ContactInfo{})
 var (
 	errContactInfoExists           = errors.DefineAlreadyExists("contact_info_exists", "contact info already exists")
 	errMatchingContactInfoNotFound = errors.DefineAlreadyExists("contact_info_not_found", "matching contact info not found")
+	errNoReference                 = errors.DefineInvalidArgument("missing_ref", "missing reference")
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/contact_info.go]: - "no" vs "missing"
- "ref" vs "reference"

The reason why we need full consistency is to trace errors back to where they're returned.

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 15]
```diff
@@ -109,6 +109,8 @@ var contactInfoFlags = util.FieldFlags(&ttnpb.ContactInfo{})
...
+	errNoToken                     = errors.DefineInvalidArgument("missing_token", "missing validation token")
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/contact_info.go]: "no" vs "missing"

Comment (johanstokking) [reply cmd/ttn-lw-cli/commands/contact_info.go]: Also beware that these errors are defined in `cmd/ttn-lw-cli/commands`, so they should be somewhat specific. Token and reference can mean different things in different contexts, so:

```go
errNoValidationToken = errors.DefineInvalidArgument("no_validation_token", "no validation token")
```

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 16]
```diff
@@ -173,8 +175,37 @@ func contactInfoCommands(entity string, getID func(cmd *cobra.Command, args []st
...
+			}
+			return io.Write(os.Stdout, config.OutputFormat, res)
+		},
+	}
+	validate.Flags().String("reference", "", "reference of your contact info validation (received by email)")
+	validate.Flags().String("token", "", "validation token (received by email)")
 	add.Flags().AddFlagSet(contactInfoFlags)
 	remove.Flags().AddFlagSet(contactInfoFlags)
 	cmd.AddCommand(add, remove)
+	cmd.AddCommand(validate)
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/contact_info.go]: Move this under defining the flags of `validate`, i.e. move to L206

File: cmd/ttn-lw-cli/commands/contact_info.go [hunk 17]
```diff
@@ -173,6 +175,35 @@ func contactInfoCommands(entity string, getID func(cmd *cobra.Command, args []st
...
+		Short: "Validate contact info",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			reference, _ := cmd.Flags().GetString("reference")
+			if reference == "" {
+				return errNoReference
+			}
+			token, _ := cmd.Flags().GetString("token")
+			if token == "" {
+				return errNoValidationToken
+			}
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/contact_info.go]: @htdvisser if these are always required, shouldn't they be taken from `args`?

Comment (htdvisser) [reply cmd/ttn-lw-cli/commands/contact_info.go]: Already commented that in a previous review: https://github.com/TheThingsNetwork/lorawan-stack/pull/1533#discussion_r344237042

**Pull request merged**